third_party/move/move-model/src/ty.rs

```diff
@@ -149,8 +149,9 @@ pub enum Constraint {
     /// The type must not be a phantom type. A phantom type is only allowed
     /// as a type argument for a phantom type parameter.
     NoPhantom,
-    /// The type must have the given set of abilities.
-    HasAbilities(AbilitySet),
+    /// The type must have the given set of abilities. The boolean indicates whether
+    /// type parameters are exempted.
+    HasAbilities(AbilitySet, bool),
```

Comment:
This needs more explanation, something like
```
(E.g., for type `S<T>` to have `drop`, ordinarily parameter `T` must also have `drop`,
but sometimes (???) we don't require that.)
```
But I hope you can fill in `sometimes (???)` because I don't quite get it.

Reply:
Ok, I made an enum out of this, for clarity.

Reply:
Much nicer, thanks!